docs/sphinx/manual/configuration.rst
******************
Configuration File
******************
The StarCluster configuration file uses `ini formatting
<http://en.wikipedia.org/wiki/INI_file>`_. It is made up of various sections
which are described here in detail. This document explains how to configure the
three required sections **[aws info]**, **[keypair]**, and **[cluster]** as
well as optional **[global]**, **[volume]**, **[permission]**, and **[plugin]**
sections.

Creating the configuration file
-------------------------------
The default starcluster config lives in ``~/.starcluster/config``. You can
either create this file manually or have starcluster create it for you with an
example configuration template (recommended).

To have StarCluster generate an example configuration file at the default
config location (``~/.starcluster/config``), simply run "starcluster help" at
the command-line.  Provided that the configuration file does not exist, you
will see the following::

    $ starcluster help
    StarCluster - (http://star.mit.edu/cluster)
    Software Tools for Academics and Researchers (STAR)
    Please submit bug reports to starcluster@mit.edu

    cli.py:475 - ERROR - Config file /home/user/.starcluster/config does not exist

    Options:
    --------
    [1] Show the StarCluster config template
    [2] Write config template to /home/user/.starcluster/config
    [q] Quit

    Please enter your selection:

Selecting 1 will print the example configuration file template to standard
output.

Selecting 2 will write the configuration file template to ~/.starcluster/config

The configuration template provided by StarCluster should be ready to go
out-of-the-box after filling in your Amazon Web Services credentials and
setting up a keypair. This example config provides a simple *cluster template*
called ``smallcluster`` that is set as the default *cluster template*.

Storing the config in an alternate location
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
If you wish to store your StarCluster config in a location other than the
default (``~/.starcluster/config``), you will need to set the
``STARCLUSTER_CONFIG`` environment variable to point to your file::

    $ export STARCLUSTER_CONFIG="/path/to/starcluster/config"

After doing so, all StarCluster commands will use the config identified by
``STARCLUSTER_CONFIG``.

Alternatively, you can specify the global ``--config`` (``-c``) option with
every StarCluster command you use. For example::

    $ starcluster -c /path/to/starcluster/config listclusters

In either case, if the config didn't exist at the specified path you would be
prompted with the same menu above offering to generate a template at the
specified path::

    $ starcluster -c /path/to/nonexistent/config listclusters
    StarCluster - (http://star.mit.edu/cluster)
    Software Tools for Academics and Researchers (STAR)
    Please submit bug reports to starcluster@mit.edu

    cli.py:475 - ERROR - Config file /path/to/nonexistent/config does not exist

    Options:
    --------
    [1] Show the StarCluster config template
    [2] Write config template to /path/to/nonexistent/config
    [q] Quit

    Please enter your selection:

Loading the config from the web
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The config can also be loaded from a web url, however, if you choose to do so
you should be *very* careful not to publicly host AWS credentials, keys, and
other private information::

    $ starcluster -c http://localhost/starcluster.cfg listclusters

.. seealso::

    See also: :ref:`splitting-the-config`

Amazon Web Services Credentials
-------------------------------
The first required section in the configuration file is **[aws info]**. This
section specifies all of your AWS credentials. The following settings are
required:

.. code-block:: ini

    [aws info]
    # replace these with your AWS keys (required)
    aws_access_key_id = #your_aws_access_key_id
    aws_secret_access_key = #your_secret_access_key
    # these settings are optional and only used for creating new AMIs
    aws_user_id= #your userid
    ec2_cert = /path/to/your/ec2_cert.pem
    ec2_private_key = /path/to/your/ec2_pk.pem

All of the settings in the **[aws info]** section can be overridden by the
environment. StarCluster will log a warning whenever it uses settings from the
environment. For example::

   $ export AWS_ACCESS_KEY_ID=your_aws_access_key_id
   $ export AWS_SECRET_ACCESS_KEY=your_secret_access_key
   $ starcluster listclusters
   *** WARNING - Setting 'AWS_SECRET_ACCESS_KEY' from environment...
   *** WARNING - Setting 'AWS_ACCESS_KEY_ID' from environment...

Amazon EC2 Regions
------------------
StarCluster uses the us-east-1 EC2 region by default. If you wish to
permanently use a different EC2 region you will need to specify the following
additional settings in your **[aws info]** section:

.. code-block:: ini

    [aws info]
    aws_region_name = eu-west-1
    aws_region_host = ec2.eu-west-1.amazonaws.com

Here ``aws_region_name`` is the name of the region you wish to use and
``aws_region_host`` is the region-specific EC2 endpoint host. Below is a table of
EC2 region-specific endpoints:

=====================  ==================================
aws_region_name        aws_region_host
=====================  ==================================
us-east-1              ec2.us-east-1.amazonaws.com
us-west-1              ec2.us-west-1.amazonaws.com
eu-west-1              ec2.eu-west-1.amazonaws.com
ap-southeast-1         ec2.ap-southeast-1.amazonaws.com
ap-northeast-1         ec2.ap-northeast-1.amazonaws.com
=====================  ==================================

.. _list from Amazon: http://docs.amazonwebservices.com/general/latest/gr/rande.html#ec2_region

The above table is only for convenience. In practice you should use the
official `list from Amazon`_ instead.

Switching Regions via Command Line
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
StarCluster also supports quickly switching between EC2 regions via the command
line without having to change your config. To switch regions at the command
line use the global *-r* (*--region*) option::

    $ starcluster -r us-west-1 listpublic

The above example runs the **listpublic** command in the ``us-west-1`` region.
Similarly, you will need to pass the global *-r* option to all of StarCluster's
commands in order to switch regions via the command line.

.. seealso::

    See also: :ref:`tips-for-switching-regions`

Amazon S3 Region-Specific Endpoints
-----------------------------------
.. _amazon: http://aws.amazon.com/articles/3912
.. note::

   Switching S3 endpoints is usually not necessary. From amazon_: Switching to
   a region-specific S3 endpoint is completely optional.  The main advantage of
   doing so is to reduce the temporary latency you might experience immediately
   after creating a bucket in a specific region.  This temporary latency
   typically lasts less than one hour.

StarCluster uses s3.amazonaws.com as the S3 endpoint by default. If you'd like
to switch S3 endpoints you can do so by specifying an additional
``aws_s3_host`` setting in your **[aws info]** section:

.. code-block:: ini

    [aws info]
    aws_region_name = us-west-1
    aws_region_name = ec2.us-west-1.amazonaws.com
    aws_s3_host = s3-us-west-1.amazonaws.com

Below is a table of S3 region-specific endpoints:

================  =================================
Region            aws_s3_host
================  =================================
us-east-1         s3.amazonaws.com
us-west-1         s3-us-west-1.amazonaws.com
eu-west-1         s3-eu-west-1.amazonaws.com
ap-southeast-1    s3-ap-southeast-1.amazonaws.com
ap-northeast-1    s3-ap-northeast-1.amazonaws.com
================  =================================

.. _proxy-config:

Using a Proxy Host
------------------
StarCluster can also be configured to use a proxy host when connecting to AWS
by specifying the following settings in your **[aws info]** section:

**aws_proxy** - The name of the proxy host to use for connecting to AWS.

**aws_proxy_port** - The port number to use to connect to the proxy host.

**aws_proxy_user** - The user name to use when authenticating with proxy host.

**aws_proxy_pass** - The password to use when authenticating with proxy host.

.. _boto: http://github.com/boto/boto

StarCluster will use the settings above when creating the `boto`_ connection
object used to communicate with AWS. Example:

.. code-block:: ini

   [aws info]
   aws_proxy = your.proxyhost.com
   aws_proxy_port = 8080
   aws_proxy_user = yourproxyuser
   aws_proxy_pass = yourproxypass

Amazon EC2 Keypairs
-------------------
In addition to supplying your **[aws info]** you must also define at least one
**[keypair]** section that represents one of your keypairs on Amazon EC2.
Amazon EC2 keypairs are used by StarCluster to connect and configure your
instances.

You should define a new **[keypair]** section for each Amazon EC2 keypair you
wish to use with StarCluster.  As an example, suppose we have two keypairs on
Amazon EC2 that we wish to use with StarCluster named ``mykeypair1`` and
``mykeypair2`` on Amazon.

.. note::

   If you do not know the name of your keypair(s), use StarCluster's
   **listkeypairs** command to obtain a list of your current EC2 keypairs. The
   **[keypair]** section name *must* match the name of the keypair on Amazon
   EC2.

To configure StarCluster for these keypairs we define a **[keypair]** section
for each of them in the configuration file:

.. code-block:: ini

    [keypair mykeypair1]
    # this is the path to your openssh private key for mykeypair4
    key_location=/path/to/your/mykeypair1.rsa

    [keypair mykeypair3]
    # this is the path to your openssh private key for mykeypair2
    key_location=/path/to/your/mykeypair2.rsa

These keypair sections can now be referenced in a *cluster template's*
**keyname** setting as we'll :ref:`show below <defining-cluster-templates>` in
an example *cluster template*.

.. _AWS web console: http://aws.amazon.com/console

.. note::

   In order for StarCluster to interact with *any* instances you have on EC2,
   the keypair used to launch those instances *must* be defined in the
   config. You can check what keypairs were used to launch an instance using
   StarCluster's **listinstances** command or the `AWS web console`_.

.. _defining-cluster-templates:

Defining Cluster Templates
--------------------------
In order to launch StarCluster(s) on Amazon EC2, you must first provide a
*cluster template* that contains all of the configuration for the cluster. A
*cluster template* is simply a **[cluster]** section in the config. Once a
*cluster template* has been defined, you can launch multiple StarClusters from
it. Below is an example *cluster template* called ``smallcluster`` which
defines a 2-node cluster using ``m1.small`` EC2 instances and the mykeypair1
keypair we defined above:

.. code-block:: ini

    [cluster smallcluster]
    keyname = mykeypair1
    cluster_size = 2
    cluster_user = sgeadmin
    cluster_shell = bash
    master_image_id = ami-0330d16a
    master_instance_type = m1.small
    node_image_id = ami-0330d16a
    node_instance_type = m1.small

Cluster Settings
^^^^^^^^^^^^^^^^
The table below describes all required and optional settings for a cluster
template in detail.

+----------------------+----------+---------------------------------------------------------------------------------+
| Setting              | Required | Description                                                                     |
+======================+==========+=================================================================================+
| keyname              | **Yes**  | The keypair to use for the cluster (the keypair must be defined in a            |
|                      |          | **[keypair]** section)                                                          |
+----------------------+----------+---------------------------------------------------------------------------------+
| cluster_size         | **Yes**  | Number of nodes in the cluster (including master)                               |
+----------------------+----------+---------------------------------------------------------------------------------+
| node_image_id        | **Yes**  | The AMI to use for worker nodes                                                 |
+----------------------+----------+---------------------------------------------------------------------------------+
| node_instance_type   | **Yes**  | The instance type for worker nodes                                              |
+----------------------+----------+---------------------------------------------------------------------------------+
| cluster_user         | No       | The cluster user to create (defaults to sgeadmin)                               |
+----------------------+----------+---------------------------------------------------------------------------------+
| cluster_shell        | No       | Sets the cluster user's shell (default: bash, options: bash, zsh, csh, ksh,     |
|                      |          | tcsh)                                                                           |
+----------------------+----------+---------------------------------------------------------------------------------+
| dns_prefix           | No       | If True, prefixes the dns name of nodes with the cluster tag. For example:      |
|                      |          | master --> mycluster-master                                                     |
+----------------------+----------+---------------------------------------------------------------------------------+
| master_image_id      | No       | The AMI to use for the master node. (defaults to **node_image_id**)             |
+----------------------+----------+---------------------------------------------------------------------------------+
| master_instance_type | No       | The instance type for the master node. (defaults to **node_instance_type**)     |
+----------------------+----------+---------------------------------------------------------------------------------+
| userdata_scripts     | No       | List of userdata scripts to use when launching instances                        |
+----------------------+----------+---------------------------------------------------------------------------------+
| volumes              | No       | List of EBS volumes to attach and NFS-share to the cluster (each volume must be |
|                      |          | defined in a **[volume]** section)                                              |
+----------------------+----------+---------------------------------------------------------------------------------+
| plugins              | No       | List of StarCluster plugins to use when launching the cluster (each plugin must |
|                      |          | be defined in a **[plugin]** section)                                           |
+----------------------+----------+---------------------------------------------------------------------------------+
| permissions          | No       | List of permissions to apply to the cluster's security group (each permission   |
|                      |          | must be defined in a **[permission]** section)                                  |
+----------------------+----------+---------------------------------------------------------------------------------+
| userdata_scripts     | No       | List of user data scripts to run on boot for each instance in the cluster       |
+----------------------+----------+---------------------------------------------------------------------------------+
| spot_bid             | No       | Always use spot instances with this cluster template                            |
+----------------------+----------+---------------------------------------------------------------------------------+
| force_spot_master    | No       | When requesting a spot cluster this setting forces the master node to also be a |
|                      |          | spot instance (default is for master not to be a spot instance for stability)   |
+----------------------+----------+---------------------------------------------------------------------------------+
| availability_zone    | No       | Launch all cluster instances in a single availability zone (defaults to any     |
|                      |          | available zone)                                                                 |
+----------------------+----------+---------------------------------------------------------------------------------+
| disable_queue        | No       | Disables the setup and configuration of the Open Grid Scheduler (OGS, formerly  |
|                      |          | SGE)                                                                            |
+----------------------+----------+---------------------------------------------------------------------------------+
| disable_cloudinit    | No       | Do not use cloudinit for cluster accounting (only required if using             |
|                      |          | non-cloudinit enabled AMIs)                                                     |
+----------------------+----------+---------------------------------------------------------------------------------+
| subnet_id            | No       | The VPC subnet to use when launching cluster instances                          |
+----------------------+----------+---------------------------------------------------------------------------------+
| public_ips           | No       | Automatically assign a public IP address to all VPC cluster instances. Set to   |
|                      |          | False to disable public IP addresses from being assigned to your VPC cluster    |
|                      |          | instances. **WARNING**: If set to False StarCluster must be used from a machine |
|                      |          | within the VPC in order to connect to the cluster. Default is `True`.           |
+----------------------+----------+---------------------------------------------------------------------------------+

.. _using-vpc:

Using the Virtual Private Cloud (VPC)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
From Amazon's `VPC page <http://aws.amazon.com/vpc/>`_:

    "Amazon Virtual Private Cloud (Amazon VPC) lets you provision a logically
    isolated section of the Amazon Web Services (AWS) Cloud where you can
    launch AWS resources in a virtual network that you define. You have
    complete control over your virtual networking environment, including
    selection of your own IP address range, creation of subnets, and
    configuration of route tables and network gateways."

New AWS accounts use VPC by default and StarCluster supports this configuration
without user intervention. However, users with `custom` configured VPCs need to
pay attention to two additional cluster settings: **subnet_id** and
**public_ips**. In order to launch your clusters within a specific VPC you will
at least need to specify the VPC subnet id in your config:

.. code-block:: ini

    [cluster smallcluster]
    keyname = mykeypair1
    cluster_size = 2
    node_image_id = ami-0330d16a
    node_instance_type = m1.small
    subnet_id = subnet-99999999

Alternatively, you can specify (or override) the subnet id at runtime via the
``--subnet-id`` (``-N``) option to the ``start`` command::

    $ starcluster start -N subnet-88888888 mycluster

By default StarCluster automatically assigns a public IP address to all VPC
cluster instances so that it can connect from anywhere. If you prefer not to be
able to connect to the cluster outside of the VPC you should also set
**public_ips** to ``False`` in your cluster template:

.. warning::

    Disabling public IPs means that StarCluster must be used from a machine
    within the VPC in order to connect to the cluster.

.. code-block:: ini

    [cluster smallcluster]
    keyname = mykeypair1
    cluster_size = 2
    node_image_id = ami-0330d16a
    node_instance_type = m1.small
    subnet_id = subnet-99999999
    public_ips = False

This will disable automatically assigning an IP address to all cluster
instances and instead rely on your custom VPC network configuration. You can
also disable public ips using the ``--no-public-ips`` option to the
``start`` command::

    $ starcluster start -N subnet-88888888 --no-public-ips mycluster

.. note::

    The ``--no-public-ips`` option only applies to VPC clusters - this feature
    is not available for non-VPC clusters.

Defining Multiple Cluster Templates
-----------------------------------
You are not limited to defining just one *cluster template*. StarCluster allows
you to define multiple independent cluster templates by simply creating a new
**[cluster]** section with all of the same settings (different values of
course).

However, you may find that defining new *cluster templates* is some what
repetitive with respect to redefining the same settings over and over. To
remedy this situation, StarCluster allows *cluster templates* to extend other
*cluster templates*:

.. code-block:: ini

    [cluster mediumcluster]
    # Declares that this cluster uses smallcluster's settings as defaults
    extends = smallcluster
    # this rest of this section is identical to smallcluster except for the following settings:
    keyname = mykeypair2
    node_instance_type = c1.xlarge
    cluster_size = 8
    volumes = biodata2


starcluster/templates/config.py
# Copyright 2009-2013 Justin Riley
#
# This file is part of StarCluster.
#
# StarCluster is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# StarCluster is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with StarCluster. If not, see <http://www.gnu.org/licenses/>.

from starcluster import static

config_template = """\
####################################
## StarCluster Configuration File ##
####################################
[global]
# Configure the default cluster template to use when starting a cluster
# defaults to 'smallcluster' defined below. This template should be usable
# out-of-the-box provided you've configured your keypair correctly
DEFAULT_TEMPLATE=smallcluster
# enable experimental features for this release
#ENABLE_EXPERIMENTAL=True
# number of seconds to wait when polling instances (default: 30s)
#REFRESH_INTERVAL=15
# specify a web browser to launch when viewing spot history plots
#WEB_BROWSER=chromium
# split the config into multiple files
#INCLUDE=~/.starcluster/aws, ~/.starcluster/keys, ~/.starcluster/vols

#############################################
## AWS Credentials and Connection Settings ##
#############################################
[aws info]
# This is the AWS credentials section (required).
# These settings apply to all clusters
# replace these with your AWS keys
AWS_ACCESS_KEY_ID = #your_aws_access_key_id
AWS_SECRET_ACCESS_KEY = #your_secret_access_key
# replace this with your account number
AWS_USER_ID= #your userid
# Uncomment to specify a different Amazon AWS region  (OPTIONAL)
# (defaults to us-east-1 if not specified)
# NOTE: AMIs have to be migrated!
#AWS_REGION_NAME = eu-west-1
#AWS_REGION_HOST = ec2.eu-west-1.amazonaws.com
# Uncomment these settings when creating an instance-store (S3) AMI (OPTIONAL)
#EC2_CERT = /path/to/your/cert-asdf0as9df092039asdfi02089.pem
#EC2_PRIVATE_KEY = /path/to/your/pk-asdfasd890f200909.pem
# Uncomment these settings to use a proxy host when connecting to AWS
#AWS_PROXY = your.proxyhost.com
#AWS_PROXY_PORT = 8080
#AWS_PROXY_USER = yourproxyuser
#AWS_PROXY_PASS = yourproxypass

###########################
## Defining EC2 Keypairs ##
###########################
# Sections starting with "key" define your keypairs. See "starcluster createkey
# --help" for instructions on how to create a new keypair. Section name should
# match your key name e.g.:
[key mykey]
KEY_LOCATION=~/.ssh/mykey.rsa

# You can of course have multiple keypair sections
# [key myotherkey]
# KEY_LOCATION=~/.ssh/myotherkey.rsa

################################
## Defining Cluster Templates ##
################################
# Sections starting with "cluster" represent a cluster template. These
# "templates" are a collection of settings that define a single cluster
# configuration and are used when creating and configuring a cluster. You can
# change which template to use when creating your cluster using the -c option
# to the start command:
#
#     $ starcluster start -c mediumcluster mycluster
#
# If a template is not specified then the template defined by DEFAULT_TEMPLATE
# in the [global] section above is used. Below is the "default" template named
# "smallcluster". You can rename it but dont forget to update the
# DEFAULT_TEMPLATE setting in the [global] section above. See the next section
# on defining multiple templates.

[cluster smallcluster]
# change this to the name of one of the keypair sections defined above
KEYNAME = mykey
# number of ec2 instances to launch
CLUSTER_SIZE = 2
# create the following user on the cluster
CLUSTER_USER = sgeadmin
# optionally specify shell (defaults to bash)
# (options: %(shells)s)
CLUSTER_SHELL = bash
# Uncomment to prepent the cluster tag to the dns name of all nodes created
# using this cluster config.  ie: mycluster-master and mycluster-node001
# If you choose to enable this option, it's recommended that you enable it in
# the DEFAULT_TEMPLATE so all nodes will automatically have the prefix
# DNS_PREFIX = True
# AMI to use for cluster nodes. These AMIs are for the us-east-1 region.
# Use the 'listpublic' command to list StarCluster AMIs in other regions
# The base i386 StarCluster AMI is %(x86_ami)s
# The base x86_64 StarCluster AMI is %(x86_64_ami)s
# The base HVM StarCluster AMI is %(hvm_ami)s
NODE_IMAGE_ID = %(x86_64_ami)s
# instance type for all cluster nodes
# (options: %(instance_types)s)
NODE_INSTANCE_TYPE = m1.small
# Launch cluster in a VPC subnet (OPTIONAL)
#SUBNET_ID=subnet-99999999
# Do not assign public IPs to cluster nodes (VPC-ONLY) (OPTIONAL)
# WARNING: You must be on a machine within the VPC in order for StarCluster to
# connect to the cluster if PUBLIC_IPS is set to False
#PUBLIC_IPS=False
# Uncomment to disable installing/configuring a queueing system on the
# cluster (SGE)
#DISABLE_QUEUE=True
# Uncomment to specify a different instance type for the master node (OPTIONAL)
# (defaults to NODE_INSTANCE_TYPE if not specified)
#MASTER_INSTANCE_TYPE = m1.small
# Uncomment to specify a separate AMI to use for the master node. (OPTIONAL)
# (defaults to NODE_IMAGE_ID if not specified)
#MASTER_IMAGE_ID = %(x86_64_ami)s (OPTIONAL)
# availability zone to launch the cluster in (OPTIONAL)
# (automatically determined based on volumes (if any) or
# selected by Amazon if not specified)
#AVAILABILITY_ZONE = us-east-1c
# list of volumes to attach to the master node (OPTIONAL)
# these volumes, if any, will be NFS shared to the worker nodes
# see "Configuring EBS Volumes" below on how to define volume sections
#VOLUMES = oceandata, biodata
# list of plugins to load after StarCluster's default setup routines (OPTIONAL)
# see "Configuring StarCluster Plugins" below on how to define plugin sections
#PLUGINS = myplugin, myplugin2
# list of permissions (or firewall rules) to apply to the cluster's security
# group (OPTIONAL).
#PERMISSIONS = ssh, http
# Add nodes to these pre-defined security groups in addition to the
# starcluster security group. (optional) 
#STATIC_SECURITY_GROUPS = default, services
# Uncomment to always create a spot cluster when creating a new cluster from
# this template. The following example will place a $0.50 bid for each spot
# request.
#SPOT_BID = 0.50
# Uncomment to specify one or more userdata scripts to use when launching
# cluster instances. Supports cloudinit. All scripts combined must be less than
# 16KB
#USERDATA_SCRIPTS = /path/to/script1, /path/to/script2

###########################################
## Defining Additional Cluster Templates ##
###########################################
# You can also define multiple cluster templates. You can either supply all
# configuration options as with smallcluster above, or create an
# EXTENDS=<cluster_name> variable in the new cluster section to use all
# settings from <cluster_name> as defaults. Below are example templates that
# use the EXTENDS feature:

# [cluster mediumcluster]
# Declares that this cluster uses smallcluster as defaults
# EXTENDS=smallcluster
# This section is the same as smallcluster except for the following settings:
# KEYNAME=myotherkey
# NODE_INSTANCE_TYPE = c1.xlarge
# CLUSTER_SIZE=8
# VOLUMES = biodata2

# [cluster largecluster]
# Declares that this cluster uses mediumcluster as defaults
# EXTENDS=mediumcluster
# This section is the same as mediumcluster except for the following variables:
# CLUSTER_SIZE=16

#############################
## Configuring EBS Volumes ##
#############################
# StarCluster can attach one or more EBS volumes to the master and then
# NFS_share these volumes to all of the worker nodes. A new [volume] section
# must be created for each EBS volume you wish to use with StarCluser. The
# section name is a tag for your volume. This tag is used in the VOLUMES
# setting of a cluster template to declare that an EBS volume is to be mounted
# and nfs shared on the cluster. (see the commented VOLUMES setting in the
# example 'smallcluster' template above) Below are some examples of defining
# and configuring EBS volumes to be used with StarCluster:

# Sections starting with "volume" define your EBS volumes
# [volume biodata]
# attach vol-c9999999 to /home on master node and NFS-shre to worker nodes
# VOLUME_ID = vol-c999999
# MOUNT_PATH = /home

# Same volume as above, but mounts to different location
# [volume biodata2]
# VOLUME_ID = vol-c999999
# MOUNT_PATH = /opt/

# Another volume example
# [volume oceandata]
# VOLUME_ID = vol-d7777777
# MOUNT_PATH = /mydata

# By default StarCluster will attempt first to mount the entire volume device,
# failing that it will try the first partition. If you have more than one
# partition you will need to set the PARTITION number, e.g.:
# [volume oceandata]
# VOLUME_ID = vol-d7777777
# MOUNT_PATH = /mydata
# PARTITION = 2

############################################
## Configuring Security Group Permissions ##
############################################
# Sections starting with "permission" define security group rules to
# automatically apply to newly created clusters. IP_PROTOCOL in the following
# examples can be can be: tcp, udp, or icmp. CIDR_IP defaults to 0.0.0.0/0 or
# "open to the # world"

# open port 80 on the cluster to the world
# [permission http]
# IP_PROTOCOL = tcp
# FROM_PORT = 80
# TO_PORT = 80

# open https on the cluster to the world
# [permission https]
# IP_PROTOCOL = tcp
# FROM_PORT = 443
# TO_PORT = 443

# open port 80 on the cluster to an ip range using CIDR_IP
# [permission http]
# IP_PROTOCOL = tcp
# FROM_PORT = 80
# TO_PORT = 80
# CIDR_IP = 18.0.0.0/8

# restrict ssh access to a single ip address (<your_ip>)
# [permission ssh]
# IP_PROTOCOL = tcp
# FROM_PORT = 22
# TO_PORT = 22
# CIDR_IP = <your_ip>/32


#####################################
## Configuring StarCluster Plugins ##
#####################################
# Sections starting with "plugin" define a custom python class which perform
# additional configurations to StarCluster's default routines. These plugins
# can be assigned to a cluster template to customize the setup procedure when
# starting a cluster from this template (see the commented PLUGINS setting in
# the 'smallcluster' template above). Below is an example of defining a user
# plugin called 'myplugin':

# [plugin myplugin]
# NOTE: myplugin module must either live in ~/.starcluster/plugins or be
# on your PYTHONPATH
# SETUP_CLASS = myplugin.SetupClass
# extra settings are passed as __init__ arguments to your plugin:
# SOME_PARAM_FOR_MY_PLUGIN = 1
# SOME_OTHER_PARAM = 2

######################
## Built-in Plugins ##
######################
# The following plugins ship with StarCluster and should work out-of-the-box.
# Uncomment as needed. Don't forget to update your PLUGINS list!
# See http://star.mit.edu/cluster/docs/latest/plugins for plugin details.
#
# Use this plugin to install one or more packages on all nodes
# [plugin pkginstaller]
# SETUP_CLASS = starcluster.plugins.pkginstaller.PackageInstaller
# # list of apt-get installable packages
# PACKAGES = mongodb, python-pymongo
#
# Use this plugin to create one or more cluster users and download all user ssh
# keys to $HOME/.starcluster/user_keys/<cluster>-<region>.tar.gz
# [plugin createusers]
# SETUP_CLASS = starcluster.plugins.users.CreateUsers
# NUM_USERS = 30
# # you can also comment out NUM_USERS and specify exact usernames, e.g.
# # usernames = linus, tux, larry
# DOWNLOAD_KEYS = True
#
# Use this plugin to configure the Condor queueing system
# [plugin condor]
# SETUP_CLASS = starcluster.plugins.condor.CondorPlugin
#
# The SGE plugin is enabled by default and not strictly required. Only use this
# if you want to tweak advanced settings in which case you should also set
# DISABLE_QUEUE=TRUE in your cluster template. See the plugin doc for more
# details.
# [plugin sge]
# SETUP_CLASS = starcluster.plugins.sge.SGEPlugin
# MASTER_IS_EXEC_HOST = False
#
# The IPCluster plugin configures a parallel IPython cluster with optional
# web notebook support. This allows you to run Python code in parallel with low
# latency message passing via ZeroMQ.
# [plugin ipcluster]
# SETUP_CLASS = starcluster.plugins.ipcluster.IPCluster
# # Enable the IPython notebook server (optional)
# ENABLE_NOTEBOOK = True
# # Set a password for the notebook for increased security
# # This is optional but *highly* recommended
# NOTEBOOK_PASSWD = a-secret-password
# # Set a custom directory for storing/loading notebooks (optional)
# NOTEBOOK_DIRECTORY = /path/to/notebook/dir
# # Set a custom packer. Must be one of 'json', 'pickle', or 'msgpack'
# # This is optional.
# PACKER = pickle
#
# Use this plugin to create a cluster SSH "dashboard" using tmux. The plugin
# creates a tmux session on the master node that automatically connects to all
# the worker nodes over SSH. Attaching to the session shows a separate window
# for each node and each window is logged into the node via SSH.
# [plugin tmux]
# SETUP_CLASS = starcluster.plugins.tmux.TmuxControlCenter
#
# Use this plugin to change the default MPI implementation on the
# cluster from OpenMPI to MPICH2.
# [plugin mpich2]
# SETUP_CLASS = starcluster.plugins.mpich2.MPICH2Setup
#
# Configure a hadoop cluster. (includes dumbo setup)
# [plugin hadoop]
# SETUP_CLASS = starcluster.plugins.hadoop.Hadoop
#
# Configure a distributed MySQL Cluster
# [plugin mysqlcluster]
# SETUP_CLASS = starcluster.plugins.mysql.MysqlCluster
# NUM_REPLICAS = 2
# DATA_MEMORY = 80M
# INDEX_MEMORY = 18M
# DUMP_FILE = test.sql
# DUMP_INTERVAL = 60
# DEDICATED_QUERY = True
# NUM_DATA_NODES = 2
#
# Install and setup an Xvfb server on each cluster node
# [plugin xvfb]
# SETUP_CLASS = starcluster.plugins.xvfb.XvfbSetup
""" % {
    'x86_ami': static.BASE_AMI_32,
    'x86_64_ami': static.BASE_AMI_64,
    'hvm_ami': static.BASE_AMI_HVM,
    'instance_types': ', '.join(static.INSTANCE_TYPES.keys()),
    'shells': ', '.join(static.AVAILABLE_SHELLS.keys()),
}

DASHES = '-' * 10
copy_below = ' '.join([DASHES, 'COPY BELOW THIS LINE', DASHES])
end_copy = ' '.join([DASHES, 'END COPY', DASHES])
copy_paste_template = '\n'.join([copy_below, config_template, end_copy]) + '\n'
